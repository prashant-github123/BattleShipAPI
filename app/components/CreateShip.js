import React, { Component } from 'react';
class CreateShip extends Component {
  constructor(props) {
    super(props);
    this.state = {
      data: '',
      rows: [0, 1, 2, 3, 4, 5, 6],
      cols: [0, 1, 2, 3, 4, 5, 6],
      shipSelected: false,
      selectable: false,
      selectedRow: null,
      selectedCol: null,
      horizontal: true,
      shipSize: 3,
      selectedCods: [],
    };
  }
  onShipSelect = () => {
    this.setState({
      horizontal: !this.state.horizontal,
    });
  }
  componentWillMount() {
    const {boardSize} = this.props.currentUser;
    let temp=[];
    for(let i=0; i<boardSize; i++){
      temp.push(i)
    }
    this.setState({
      rows: temp,
      cols: temp
    });
  }
  /* highlight cells user put the mouse*/
  mouseOver = (e) => {
    const { horizontal, shipSize, rows, cols } = this.state;
    const val = (e.target.id).split('');
    this.setState({
      selectedRow: parseInt(val[0], 10),
      selectedCol: parseInt(val[1], 10),
    });
    if (horizontal) {
      if (parseInt(val[1], 10) <= (cols.length - shipSize)) {
        this.setState({ selectable: true });
      } else {
        this.setState({ selectable: false });
      }
    } else if (parseInt(val[0], 10) <= (rows.length - shipSize)) {
      this.setState({ selectable: true });
    } else {
      this.setState({ selectable: false });
    }
  }
  /*
  * Handle Cell Click
  * Set the co-oridnates to state / browser session
  */
  handleClick = (e) => {
    const val = parseInt(e.target.id, 10);
    const cods = [];
    const { horizontal, selectable, shipSize } = this.state;

    if (horizontal && selectable) {
      for (let i = 0; i < shipSize; i += 1) {
        cods.push(val + i);
      }
    } else if (selectable) {
      for (let i = 0; i < shipSize; i += 1) {
        cods.push(val + (i * 10));
      }
    }
    this.setState({
      selectedCods: cods,
      shipSelected: true,
    });
    sessionStorage.setItem('cords', cods);
  }

  /*
  * Handle Next  Button Click
  * Send the Co-ordinates to Micro Service
  */
  handleNextClick = () => {
    const { currentUser, placeShip } = this.props;
    const { selectedCods } = this.state;
    placeShip(currentUser, selectedCods);
  }
  /*
  * Vertical Ship Placement
  */
  verticalSelection = (row) => {
    const { selectable, selectedRow, shipSize } = this.state;
    if (selectable && ((row >= selectedRow) && (row <= (selectedRow + (shipSize - 1))))) {
      return 'shipSelected';
    } else if ((selectedRow <= row) && !selectable) {
      return 'shipUnselected';
    }
    return '';
  }
  /*
  * Horizontal Ship Placement
  */
  horizontalSelection = (col) => {
    const { selectable, selectedCol, shipSize } = this.state;
    if (selectable && ((col >= selectedCol) && (col <= (selectedCol + (shipSize - 1))))) {
      return 'shipSelected';
    } else if ((selectedCol <= col) && !selectable) {
      return 'shipUnselected';
    }
    return '';
  }
  /*
  * Create Cells
  */
  cellDivs = (row) => {
    const { selectedRow, horizontal, selectedCol, shipSelected, shipSize, selectedCods, cols } = this.state;
    let selectedTabs = '';

    if (row === selectedRow && horizontal && !shipSelected) {
      return (
        cols.map((col) => {
          selectedTabs = this.horizontalSelection(col);
          return (
            <div
              onFocus={() => undefined}
              role="presentation"
              className={`indent ${selectedTabs}`}
              id={`${row}${col}`}
              onMouseOver={this.mouseOver}
              key={`${row}${col}`}
              onClick={this.handleClick}
            />
          );
        })
      );
    } else if (!horizontal && !shipSelected) {
      return (
        cols.map((col) => {
          if (selectedCol === col) {
            selectedTabs = this.verticalSelection(row);
          } else {
            selectedTabs = '';
          }
          return (
            <div
              onFocus={() => undefined}
              role="presentation"
              className={`indent ${selectedTabs}`}
              id={`${row}${col}`}
              onMouseOver={this.mouseOver}
              key={`${row}${col}`}
              onClick={this.handleClick}
            />
          );
        })
      );
    }
    return (
      cols.map((col) => {
        const cord = `${row}${col}`;
        let tabs = '';
        for (let i = 0; i < shipSize; i += 1) {
          if (cord == selectedCods[i]) { // eslint-disable-line eqeqeq
            tabs = 'shipSelected';
          }
        }

        return (
          <div
            onFocus={() => undefined}
            role="presentation"
            className={`indent ${tabs}`}
            id={`${row}${col}`}
            onMouseOver={this.mouseOver}
            key={`${row}${col}`}
            onClick={this.handleClick}
          />
        );
      })
    );
  }


  render() {
    const { rows, shipSelected, horizontal } = this.state;
    const tabSelection = horizontal ? 'Horizontal' : 'Vertical';
    const shipColor = shipSelected ? 'shipSelected' : 'shipUnselected';
    return (
      <div className="App">
        <div style={{ textAlign: 'center' }}>
          {
            rows.map((row) => (
              <div className={'row'} key={row}>
                {this.cellDivs(row)}
              </div>
            ))
          }
          <button className="next-submit" onClick={this.handleNextClick}> Next </button>
          <div role="presentation" className={`ship ${shipColor}`} onClick={this.onShipSelect}>{tabSelection}</div>
        </div>
      </div>
    );
  }
}
export default CreateShip;

CreateShip.propTypes = {
  placeShip: React.PropTypes.func, // eslint-disable-line
};
