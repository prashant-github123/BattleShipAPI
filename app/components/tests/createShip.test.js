import React from 'react';
import { shallow } from 'enzyme';
// import renderer from 'react-test-renderer';
import CreateShip from '../CreateShip';
// import {retriveShipLocation } from './../actions';
describe('CreateShip (Shallow passing the ship coords directly)', () => {
  let wrapper;
  let output;
  const currentUser = {
    'message':'',
    'status':true,
    'statusCode':200,
    'gameId':'',
    'playerId':'',
    'playerName':'',
  };

  beforeEach(() => {
    wrapper = shallow(<CreateShip currentUser={currentUser} />);
  });
  it('+++ render the DUMB component', () => {
    expect(wrapper.length).toEqual(1);
  });
  it('+++ contains button with class="next-submit"', () => {
    expect(wrapper.find('button.next-submit').type()).toEqual('button');
    wrapper.find('button.next-submit').simulate('click');
  });
  it('on click of box change class', () => {
    wrapper.setState({
      rows: [0, 1, 2, 3, 4, 5, 6],
      cols: [0, 1, 2, 3, 4, 5, 6],
      horizontal: true,
      shipSize: 3,
    });
    wrapper.find('.indent').slice(1).at(1).simulate('mouseOver', { target: {
      id: '0',
    } });
    // expect(wrapper.find('.indent').hasClass('shipSelected')).to.equal(true);
    wrapper.find('.indent').forEach((node) => {
      expect(node.hasClass('shipSelected')).to.equal(true);
    });
  });
  it('on click of box change class', () => {
    wrapper.setState({ 
      rows: [0, 1, 2, 3, 4, 5, 6],
      cols: [0, 1, 2, 3, 4, 5, 6],
      horizontal: true,
      shipSize: 3,
    });
    wrapper.find('.indent').slice(1).at(1).simulate('click', { target: {
      id: '0',
    } });
    wrapper.find('.indent').forEach((node) => {
      expect(node.hasClass('shipSelected')).to.equal(true);
    });
    // expect(wrapper.find('.indent').hasClass('shipSelected')).to.equal(3);
  });
});
