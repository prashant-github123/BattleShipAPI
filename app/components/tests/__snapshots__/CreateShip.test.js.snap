// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateShip --- Snapshot capturing Snapshot of CreateShip 1`] = `
<div
  className="App"
>
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    <button
      className="next-submit"
      onClick={[Function]}
    >
       Next
    </button>
    <div
      className="ship shipUnselected"
      onClick={[Function]}
      role="presentation"
    >
      Horizontal
    </div>
  </div>
</div>
`;
